### DISCUSSION TITLE
Matrix fn with numpy

### DISCUSSION THREAD

**Discussion by survuvi** [modular/modular#267] [Q&A] [answered]
I don't see why this error:

```python

from PythonInterface import Python

fn create_matrix(rows : Int, cols : Int) -> object:
    # This is equivalent to Python's `import numpy as np`
    let np = Python.import_module("numpy")

    # Create matrix with numpy:
    let matrix = np.arange(1, rows*cols+1).reshape(rows, cols)

    return matrix

# Create matrix 9 rows and 5 cols:
print(create_matrix(9, 5))

```

_error: Expression [18]:7:34: cannot call function that may raise in a context that cannot raise
    let np = Python.import_module("numpy")_

**Comment by lattner** [comment, marked as answer]
The issue here is that all calls through imported python APIs can raise an error at any time, but a 'fn' defaults not not being able to raise.  Try switching create_matrix to being a def, or mark it "raises"

**Reply by daydun** [reply]
A low priority QoL could be to explicitly mention that `import_module` can raise and `create_matrix` cannot, and suggest appropriate changes to fix the error.

**Reply by czheo** [reply]
I was also bitten by this error before: https://github.com/modularml/mojo/issues/134.
As @DayDun suggested, a better error message may help developer experience.

**Reply by iamtimdavis** [reply]
@Mogball - we should fix the error string

**Reply by Mogball** [reply]
Error message should get improved in the next release

**Comment by survuvi** [comment]
Thank you @lattner I added raises and now it works: 

```python

from PythonInterface import Python

fn create_matrix(rows : Int, cols : Int) raises -> PythonObject:
    # This is equivalent to Python's `import numpy as np`
    let np = Python.import_module("numpy")

    # Create matrix with numpy:
    let matrix = np.arange(1, rows*cols+1).reshape(rows, cols)

    return matrix

# Create matrix 9 rows and 5 cols:
print(create_matrix(9, 5))

```

**Comment by gautam-e** [comment]
I had a similar issue on a `fn` that I wanted to vectorise. 

Am I right to say that `vectorize`-ing such a `fn` is not possible because `vectorize` requires a `func: fn[Int](Int) capturing -> None` and it will not work with a `fn[Int](Int) raises capturing -> None`?

**Comment by survuvi** [comment]
I solved all my problems abandoning mojo for c++, mojo is undoubtedly an interesting project, even if it's a soup of multiple languages, but it's impossible to follow at the moment, it's too early stage and too confusing, even input doesn't work..
